### PR TITLE
luatest_helpers: reworked cluster startup to be reproducable

### PR DESCRIPTION
**Pull request by grafin**
Makes bootstrap master on cluster:start() reproducible.
While inspecting difference in logs I've noticed that inter-server communication differs even before the test is started (after cluster:start(), and before anything else). This led me to reworking cluster:start() to start servers sequentially.

### CODE REVIEW

**Comment by grafin** [comment]
Thanks @sergepetrenko for helping to dig into this issue. This PR doesn't solve #6754. But IMHO it is useful on its own, because it ensures bootstrap master is consistent on cluster build (the first added server to cluster becomes bootstrap master), and it makes tests tests more reproducable.

File: test/luatest_helpers/cluster.lua
Comment (sergepetrenko) [inline test/luatest_helpers/cluster.lua:88]: I guess this should go under the `if replication_settings[i] ~= nil`

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Thanks for the patch!
It does what it's supposed to, and I like the idea.
I only have one doubt. Are you sure that every `cluster:start` will create a brand new cluster?
What if a user wants to start everyone with exact config he passed?
What if he wants to connect server[3] to server[2], for example, and never connect it to server[1]?

I suppose this `startup ordering` should go under a special option. Something like `opts.ordered_startup`

File: test/luatest_helpers/cluster.lua
Comment (grafin) [reply test/luatest_helpers/cluster.lua:88]: Thanks for review! There was a bigger problem with that part of the code. I wanted that after start() has finished server.box_cfg was the same as before (as it was set by user) and box.cfg on the server itself was in the state user asked it to be. I've rewritten it, take a look please.

File: test/luatest_helpers/cluster.lua
Comment (sergepetrenko) [inline test/luatest_helpers/cluster.lua:101]: I guess you misunderstood my previous comment.
Imagine you have a cluster, where one server is running, and others are stopped. Then someone does `cluster:start()` on the cluster. This will nullify the already-running server's `box.cfg.replication`, because `replicatoin_settings[i]` will be `nil` for it.

So the proposed change is
```diff
diff --git a/test/luatest_helpers/cluster.lua b/test/luatest_helpers/cluster.lua
index 0d0390691..ed0ae8b20 100644
--- a/test/luatest_helpers/cluster.lua
+++ b/test/luatest_helpers/cluster.lua
@@ -98,8 +98,10 @@ function Cluster:start(opts)
 
     if sequential_startup then
         for i, server in ipairs(self.servers) do
-            server:box_config({replication = replication_settings[i]})
-            server.box_cfg.replication = replication_settings[i]
+            if replication_settings[i] ~= nil then
+                server:box_config({replication = replication_settings[i]})
+                server.box_cfg.replication = replication_settings[i]
+            end
         end
     end
 
```

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Thanks for the fixes!
Only one comment left, everything else looks good.

File: test/luatest_helpers/cluster.lua
Comment (grafin) [reply test/luatest_helpers/cluster.lua:101]: Ah, yes, I didn't understand the problem. I propose following patch, it will work both for starting cluster with already started servers (will not touch them) and for starting server where box_cfg.replication == nil (will still set it back to nil after starting)

```diff
diff --git a/test/luatest_helpers/cluster.lua b/test/luatest_helpers/cluster.lua
index 0d0390691..0be7e0ce4 100644
--- a/test/luatest_helpers/cluster.lua
+++ b/test/luatest_helpers/cluster.lua
@@ -85,6 +85,7 @@ function Cluster:start(opts)
     end

     local replication_settings = {}
+    local started_server = {}

     for i, server in ipairs(self.servers) do
         if not server.process then
@@ -93,13 +94,16 @@ function Cluster:start(opts)
                 server.box_cfg.replication = self.servers[1].net_box_uri
             end
             server:start({wait_for_readiness = sequential_startup})
+            started_server[i] = true
         end
     end

     if sequential_startup then
         for i, server in ipairs(self.servers) do
-            server:box_config({replication = replication_settings[i]})
-            server.box_cfg.replication = replication_settings[i]
+            if started_server[i] == true then
+                server:box_config({replication = replication_settings[i]})
+                server.box_cfg.replication = replication_settings[i]
+            end
         end
     end
```

Comment (sergepetrenko) [reply test/luatest_helpers/cluster.lua:101]: Looks good, thanks! I missed the other part of the problem.

Comment (grafin) [reply test/luatest_helpers/cluster.lua]: Thanks for review! I've squashed this patch and forced pushed PR branch.

**Review by sergepetrenko — COMMENTED** [review]
Thanks! The second commit LGTM. We can't merge it though until box_promote tests are merged.
I suggest you base box_promote() tests on top of this patch. Then we'll be able to merge it once Vlad gives an ack. Sorry I haven't thought of it right away.

File: test/luatest_helpers/cluster.lua
Comment (Gerold103) [inline test/luatest_helpers/cluster.lua:72]: I think the patch isn't really needed. Because in a test you either should not rely on who is the master (boostrap leader) and should find it dynamically, or if you really want one node to be the master - you can give it `read_only = false` and to the others - `read_only = true`. You could do that here too by the way. Probably it could be much simpler and would work faster because you can start all nodes at once. Also you can use non-random UUIDs - give minimal UUID to the node which needs to be the bootstrap master (voting for boot leader uses UUID comparison). It does not need to be sequential if you want just to have a specific node as a master.

This sequential startup I am sure won't really fix the flakyness. At least I can't remember when not knowing exact boot master was the reason for a fail. Usually fail happens after bootstrap in something not related to that. But I do believe this thing makes tests notably slower which is quite bad, and while in theory it could make tests probably more stable, it will come at a cost of smaller coverage.

And finally the API looks a bit strange. You assume the master is the first instance. I would rather expect it to be more explicit. Like a flag for the server which you want to be the master.

File: test/luatest_helpers/server.lua
Comment (Gerold103) [inline test/luatest_helpers/server.lua:85]: It can be simpler: `if not next(self.box_cfg) then ...`. Why do you need `rawequal()`?

**Review by Gerold103 — CHANGES_REQUESTED** [review]
While the second commit is no harm because it is not enabled by default, I still would want you to consider some alternatives. At least because the test this is going to be used for is long, and the given patch will make it even longer.

Consider alternatives which don't require this patch - `read_only` option and specific UUIDs to control who will be the boot leader.

By the way, why do you need to know who is the boot leader? Does it matter really in your main patch?

File: test/luatest_helpers/server.lua
Comment (grafin) [reply test/luatest_helpers/server.lua:85]: Just in case someone overloaded __eq method for self.box_cfg.

File: test/luatest_helpers/cluster.lua
Comment (grafin) [reply test/luatest_helpers/cluster.lua:72]: Thanks for in depth review! Setting only one server in cluster to rw and others to ro does the trick with less effort and faster! It doesn't make the test completely reproducible, but makes it fail a lot more often which helps debugging the real fix. Closing this PR now.

**Comment by grafin** [comment]
Setting read_only = true on all servers, except the desired bootstrap leader has the same effect, simpler, works faster and overall a better solution.

**Pull request closed without merging**
